polytracker/include/taintdag/fntrace.h

```diff
@@ -11,34 +11,24 @@
 #include "taintdag/fnmapping.h"
 
 #include <cstdint>
-#include <mutex>
 
 namespace taintdag {
 
-class FnTrace {
+struct Event {
 public:
-  using offset_t = uint32_t;
-  using fn_index_t = FnMapping::index_t;
-
-  struct event_t {
-    enum class kind_t : uint8_t { entry, exit };
-    kind_t kind;
-    fn_index_t function;
-  };
-
-  FnTrace(char *begin, char *end);
-  void log_fn_event(event_t::kind_t kind, fn_index_t idx);
-  size_t get_event_count();
-
-private:
-  // Trace markers
-  char *trace_begin{nullptr};
-  char *trace_end{nullptr};
-  char *events_end{nullptr};
-  // Trace mutex
-  std::mutex memory_m;
-  // Helpers
-  std::optional<offset_t> write_event(event_t event);
+  enum class kind_t : uint8_t { entry, exit };
+  kind_t kind;
+  Functions::index_t function;
+};
+
+class Events : public FixedSizeAlloc<Event> {
+public:
+  static constexpr uint8_t tag{7};
+  static constexpr size_t allocation_size{std::numeric_limits<uint32_t>::max() *
```

Comment:
In the end I decided not to do the `+1` anywhere because in the `Events` section case it overflows `allocation_size`. The `+1` is also unnecessarily confusing and overall not worth the one extra entry in any section.